pkg/kn/commands/service/service_describe.go

```diff
@@ -41,23 +125,513 @@ func NewServiceDescribeCommand(p *commands.KnParams) *cobra.Command {
+	}
+}
+
+func writeConcurrencyOptions(dw printers.PrefixWriter, desc *revisionDesc) {
+	dw.WriteColsLn(printers.LEVEL_1, "", l("Concurrency"))
+	if desc.concurrencyLimit != nil {
+		dw.WriteColsLn(printers.LEVEL_2, "", "", l("Limit"), strconv.FormatInt(*desc.concurrencyLimit, 10))
+	}
+	if desc.concurrencyTarget != nil {
+		dw.WriteColsLn(printers.LEVEL_2, "", "", l("Target"), strconv.Itoa(*desc.concurrencyTarget))
+	}
+}
+
+// ======================================================================================
+// Helper functions
+
+// Format label (extracted so that color could be added more easily to all labels)
+func l(label string) string {
+	return label + ":"
+}
+
+// Format scale in the format "min ... max" with max = ∞ if not set
+func formatScale(minScale *int, maxScale *int) string {
+	ret := "0"
+	if minScale != nil {
+		ret = strconv.Itoa(*minScale)
+	}
+
+	ret += " ... "
+
+	if maxScale != nil {
+		ret += strconv.Itoa(*maxScale)
+	} else {
+		ret += "∞"
+	}
+	return ret
+}
+
+// Format the revision name along with its generation. Use colors if enabled.
+func getRevisionNameWithGenerationAndAge(desc *revisionDesc) string {
+	return desc.name + " " +
+		"[" + strconv.Itoa(desc.configurationGeneration) + "]" +
+		" " +
+		"(" + age(desc.creationTimestamp) + ")"
+}
+
+// Used for conditions table to do own formatting for the table,
+// as the tabbed writer doesn't work nicely with colors
+func getMaxTypeLen(conditions v1beta1.Conditions) int {
+	max := 0
+	for _, condition := range conditions {
+		if len(condition.Type) > max {
+			max = len(condition.Type)
+		}
+	}
+	return max
+}
+
+// Color the type of the conditions
+func formatConditionType(condition apis.Condition) string {
+	return string(condition.Type)
+}
+
+// Status in ASCII format
+func formatStatus(status corev1.ConditionStatus) string {
+	switch status {
+	case v1.ConditionTrue:
+		return "++"
+	case v1.ConditionFalse:
+		return "--"
+	default:
+		return ""
+	}
+}
+
+// Return either image name with tag or together with its resolved digest
+func getImageDesc(desc *revisionDesc) string {
+	image := desc.image
+	if printDetails && desc.imageDigest != "" {
+		return fmt.Sprintf("%s (%s)", image, shortenDigest(desc.imageDigest))
+	}
+	return image
+}
+
+// Extract pure sha sum and shorten to 8 digits,
+// as the digest should to be user consumable. Use the resource via `kn service get`
+// to get to the full sha
+func shortenDigest(digest string) string {
+	match := imageDigestRegexp.FindStringSubmatch(digest)
+	if len(match) > 1 {
+		return string(match[1][:12])
+	}
+	return digest
+}
+
+// Write a map either compact in a single line (possibly truncated) or, if printDetails is set,
+// over multiple line, one line per key-value pair
+func writeMapDesc(dw printers.PrefixWriter, indent int, m map[string]string, label string, labelPrefix string) {
+	if len(m) == 0 {
+		return
+	}
+
+	if printDetails {
+		l := labelPrefix + label
+		for key, value := range m {
```

Comment:
I'm a little concerned that the kn prints a list of annotations and labels in random order. This is caused by looping a map. Can we put all the map keys in string slice and sort using [sort.Strings](https://golang.org/pkg/sort/#Strings) before writing them in the buffer ?

Reply:
Good catch. Let's print them in alphabetical order, will fix that.